File: bench/10_sort/bench.py
"""Benchmark 10: Bubble sort (3000 integers)."""
import time

def bubble_sort(arr):
    n = len(arr)
    for i in range(n):
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]

total = 3000
data = [(i * 31337 + 17) % 100000 for i in range(total)]

t0 = time.perf_counter()
bubble_sort(data)
ms = int((time.perf_counter() - t0) * 1000)

checksum = data[0] + data[total - 1] + data[total // 2]
ok = all(data[i] <= data[i + 1] for i in range(total - 1))
print(f"result: {checksum} sorted={ok}  time: {ms}ms")

File: bench/10_sort/bench.rs
// Benchmark 10: Bubble sort (3000 integers, pure Rust).
use std::time::Instant;

fn bubble_sort(arr: &mut Vec<i32>) {
    let n = arr.len();
    for i in 0..n {
        for j in 0..(n - i - 1) {
            if arr[j] > arr[j + 1] {
                arr.swap(j, j + 1);
            }
        }
    }
}

fn main() {
    let n = 3000usize;
    let mut data: Vec<i32> = (0..n as i32).map(|i| (i * 31337 + 17) % 100000).collect();

    let t0 = Instant::now();
    bubble_sort(&mut data);
    let ms = t0.elapsed().as_millis();

    let checksum = data[0] + data[n - 1] + data[n / 2];
    let ok = data.windows(2).all(|w| w[0] <= w[1]);
    println!("result: {checksum} sorted={ok}  time: {ms}ms");
}
